### PR TITLE
RRFS_dev1: compile DA on Jet, Hera, Orion, Wcoss

### PR DESCRIPTION
**Pull request by hu5970**
Changes for compiling RRFS DA components on JET, HERA, Orion, and WCOSS (DELL).

### CODE REVIEW

File: devbuild.sh
Comment (christinaholtNOAA) [inline devbuild.sh:78]: Why are we going away from a cmake build here? How is the environment managed?

Comment (hu5970) [reply devbuild.sh:78]: It is intent to let the GSI manage its own compile. The APP only set the run environment now.

Comment (christopherwharrop-noaa) [reply devbuild.sh:78]: I know I'm not a selected reviewer, but I get notifications and saw this come through.  I know the intent here is to "make things work", and I understand that goal and deadline driven activities.  But speaking as someone who has spent/wasted countless hours, days, weeks? wrestling with non portable build systems, trying to make them work in ways the developers did not foresee, this is the type of thing that makes life truly miserable for others.  I can say without hesitation that the lack of a proper, portable, build system is the root cause of major delays in delivering on project goals for work I've been involved with.  Compliance with the existing `cmake` build design should be a **minimum** barrier of entry for anything that goes into any part of the UFS.  Adding stuff like this takes the UFS backwards, not forwards.  I realize that what I'm asking is a heavy lift, because so many parts we want to add do not yet have a functioning `cmake` build system.  But, the price we pay when we do not adhere to standards is that users cannot do their work because they spend months trying to figure out how to work around all of the machine-dependent idiosyncrasies.  Even the existing `cmake` build still has many issues in that regard, but it is at least moving in the right direction.

Please consider what you are trading off by adding things like this "to make it work" at the expense of the standards that need to be upheld and the resulting cost to developers and users down the road. And also beware of the inertia of stuff that goes into a repository.  Once it goes in, you can't take it out as easily.  And bad things that go in, often stay bad for a very long time.  If this change is rejected, the motivation to fix it high.  If it is accepted, the motivation to do it correctly is non-existent, and will likely not happen for a very long time.

Comment (hu5970) [reply devbuild.sh:78]: @christinaholtNOAA I don't think the way I am using will cause any trouble in the future. It follows idea of separate concern. It is not UFS app's task to make GSI or other components portable or work on every platform. It is GSI developer's task.  Here, GSI has its own system to make it works on all the platforms we need. We don't need to copy everything from GSI's compiling system to UFS and maintain it in the future. If anything wrong with GSI, the GSI developer should fix it, not the UFS APP developer.

File: env/build_jet_intel.env_DA
Comment (hu5970) [reply env/build_jet_intel.env_DA:38]: Those are copied from UPP module file for compiling rrfs_util to handle grib2 files.
I just updated the rrfs_util  CMake to use the environments set by those modules. 
I don't know if hoc-stack can provide the same environment but I guess any module change needs to reset something in the CMake.

**Comment by hu5970** [comment]
@christopherwharrop-noaa @guoqing-noaa Please approve the PR if there are no issue other than unify the compile environment for GSI and FV3 model. We will follow what ever GSI developer put there for GSI run at this moment.

**Review by guoqing-noaa — CHANGES_REQUESTED** [review]
@hu5970 I think it is okay to go forward with this PR and we can make adjustments in the future if the context is somewhat changed. 

However, I would like request changes in Externals.cfg to use hashes (not branches) for both the "gsi" and "rrfs_utl" components. In the last few days, I cloned the ufs-srweather-app but failed to compile rrfs_utl because there were new updates to the "develop" branch of rrfs_utl . The new updates are consistent with this PR but not with latest ufs-srweather-app. We can avoid this kind of inconsistent issues in the future by using "hash" instead of "branch". Thanks!

File: devbuild.sh
Comment (christinaholtNOAA) [reply devbuild.sh:78]: It is the job of the UFS Apps to build all the components that it needs to run the application. It is an effort of UFS to have all those components build with CMake. Please do not undo the CMake build that was already implemented here. This change is going in the opposite direction of what we need to be accomplishing with this build.

**Review by christinaholtNOAA — DISMISSED** [review]

**Comment by hu5970** [comment]
@guoqing-noaa I have updated the Externals.cfg to use hash for GSI and rrfs_util.

File: devbuild.sh
Comment (christopherwharrop-noaa) [reply devbuild.sh:78]: I agree with @christinaholtNOAA here.  Removing the `cmake` build makes no sense to me and appears to be a regression in features/capability as it replaces a standard portable build mechanism with an ad-hoc one.  There is going in the wrong direction.  What should be happening instead is to improve upon the existing CMake such that any given package's build system does not build other external packages but uses Cmake's `find_package` utility to discover dependencies and fail if they are not found.

**Review by guoqing-noaa — APPROVED** [review]
I think the way to build GSI does not make huge differences. "ush/build.comgsi" also uses cmake and all standard UFS module files in Ming's latest GSI merge. I have no preferences on how to put it in devbuild.sh.  At current fast developing stage, either way works.

File: Externals.cfg
Comment (hu5970) [reply Externals.cfg]: Updated the GSI and put a new hash number to use the latest merged GSI.

**Comment by hu5970** [comment]
I also updated hash number for UFS_UTIL to use the lasted merged version on top of the EMC develop branch.

**Comment by hu5970** [comment]
@christopherwharrop-noaa @christinaholtNOAA 

I agree the app needs to be able to compile every component. But how to make this happen is a challenge issue and the current cmake in APP is not working well for developers. Right now, I want to APP treats the compiling of the GSI and rrfs_util as a black box. The GSI and rrfs_util is using cmake the same way as other components but with difference environment loaded. It is unnecessary at this stage to duplicate GSI and rrfs_util compiling command in devbuild.

We will keep our discussion about the best way APP should use to compile the components. I would like to suggest merge this PR to upgrade the RRFS.

**Comment by christinaholtNOAA** [comment]
@hu5970 

Merging this without approvals from the reviewers actively reviewing the PR was definitely against the protocols we've set up for getting changes into RRFS_dev1. The code change is bad practice, and we are in a situation where we are trying to share this code with a larger group. It should not be in this branch. 

I would like to revert this merge until the changes are made that are appropriate for managing this build with cmake directly.

**Comment by terraladwig** [comment]
I have a question about the difference in build options here...

If the cmake build is left as it is intended in the UFS, so using the
existing portable build mechanism, can our GSI capability still be built?
I think Chris suggested how to make this work with modifications to
isolating the external packages.  Is there significant work to make this
happen and use the existing UFS build setup?  If not, I suggest we do
this.
I think we should keep on track with the UFS capabilities as much as
possible to minimize conflicts with the community.

**Comment by hu5970** [comment]
@christinaholtNOAA I don't think you should revert his change as we need to move forward with the test. The argument we have in this PR has not meaningful impact to our realtime evaluations and it has delayed the tests needed. We can keep talking about this compiling issue but the system need to be updated. Reviewer should not block the progress of the code development and evaluation because of the argument like this.

**Pull request merged**
